README.md
# Idasen Desk Control

[Get latest version here!](https://github.com/danieldesrochers/idasen-ui/releases)

Idasen Desk control works with ikea IDÅSEN desk.


![app screenshot](./idasen-ui.png)

IDÅSEN is an electric sitting-standing desk with a Linak bluetooth controller sold by ikea.
The position of the desk can controlled by a physical switch on the desk or via bluetooth using an phone app.

This application controls the Idasen desk via bluetooth from a desktop computer and allows to store two favorite positions.

This application is built in Python on top of the IdasenDesk API made by 'newAM/idasen' which was a heavily modified fork of 'rhyst/idasen-controller'.

The Idasen desk control app (idasen-ui) has been package for Windows 10 for easier installation.


Prerequisites
=============
**The desk must be connected and paired to the computer before launching the application.**
- Open Windows Parameters
- Find your Bluetooth settings
- Put the IKEA Idasen desk in pairing mode (blue light will flash slowly)
- On the Bluetooth setting window, click "Add a Bluetooth device"
- Select "Desk 8638" and add it to your bluetooth devices

Install Idasen Desk Control
===========================
- Download latest release installer (https://github.com/danieldesrochers/idasen-ui/releases)

- Download and launch the latest installer (i.e.: idasen desk control v1.3.0.exe)

- Accept the license and choose your installation folder 

- A new 'Idasen Desk Control' shortcut will be created in Windows Start menu


How To Use Idasen Desk Control
==============================
Idasen Desk Control is a simple user interface to control your IKEA IDASEN desk.

Connect your desk
------------------
When application runs for the first time, no save configuration will be found. 
Be patient, it might take few minutes to explore all Bluetooth devices paired with the computer.
Click Bluetooth button to discover your desk and save initial configuration.

You are connected!
------------------
Once connected to the desk, current desk position will be displayed and moving buttons will be enabled. 
- Press and maintain the up/down arrow the move the desk to the desired position.
- Press M(emory) button to save the current position to 1 or 2.
- Press position 1 or 2 to move the desk to the save position.

Right-click on the Idasen Desk Control display to control the user interface
- Always on top
- Minimize to tray instead of taskbar

Known issues
============
IKEA IDASEN Desk internal Linak controller seems to have a built-in memory for previous positions. This could cause some weird move effects. The app will retry twice to move to the right direction. The built-in memory issue seems to reduce while using the application for a longer period since the previous built-in positions match those from the application.


Changelog.txt
= Release 1.3.0 =
    * Changed application installation to NSIS installer
    * Added shortcut in Windows Start Menu
    * Added uninstall option in registry settings
    * Fixed .NET Framework UAC security issue
    
= Release 1.2.1 =
    * Added minimize to tray option    
    * Added right-click menu to control tray option
    * Added minimize_to_tray setting in configuration

= Release 1.1.2 =
    * Fixed python installer issue
    * Fixed loading issue reported by some users
    * Changed log_to_file approach to overwrite log file instead of appending
    
= Release 1.1.1 =
    * Added log_to_file setting in configuration
    * Added debug trace to investigate issue reported by user.
    
= Release 1.1.0 =
    * Added right-click menu with Always-On-Top window
    * Added Always-On-Top setting in configuration
    * Improved workaround to limit bouncing desk issue when moving down.
    * Fixed UI height

= Release 1.0.0 = 
    * Initial version

idasen-ui/__init__.py
import sys
import time

from bleak import BleakClient
from bleak import discover
from typing import Dict
from typing import Optional
from typing import Tuple
import asyncio
import logging


_UUID_HEIGHT: str = "99fa0021-338a-1024-8a49-009c0215f78a"
_UUID_COMMAND: str = "99fa0002-338a-1024-8a49-009c0215f78a"
_UUID_REFERENCE_INPUT: str = "99fa0031-338a-1024-8a49-009c0215f78a"

_COMMAND_REFERENCE_INPUT_STOP: bytearray = bytearray([0x01, 0x80])
_COMMAND_UP: bytearray = bytearray([0x47, 0x00])
_COMMAND_DOWN: bytearray = bytearray([0x46, 0x00])
_COMMAND_STOP: bytearray = bytearray([0xFF, 0x00])

# height calculation offset in meters, assumed to be the same for all desks


def _bytes_to_meters(raw: bytearray) -> float:
    """ Converts a value read from the desk in bytes to meters. """
    raw_len = len(raw)
    expected_len = 4
    assert (
        raw_len == expected_len
    ), f"Expected raw value to be {expected_len} bytes long, got {raw_len} bytes"

    high_byte = int(raw[1])
    low_byte = int(raw[0])
    raw = (high_byte << 8) + low_byte
    return float(raw / 10000) + IdasenDesk.MIN_HEIGHT


class _DeskLoggingAdapter(logging.LoggerAdapter):
    """ Prepends logging messages with the desk MAC address. """

    def process(self, msg: str, kwargs: Dict[str, str]) -> Tuple[str, Dict[str, str]]:
        return f"[{self.extra['mac']}] {msg}", kwargs


class IdasenDesk:
    """
    Idasen desk.

    Args:
        mac: Bluetooth MAC address of the desk.
        exit_on_fail: If set to True, failing to connect will call ``sys.exit(1)``,
            otherwise the exception will be raised.

    Note:
        There is no locking to prevent you from running multiple movement
        coroutines simultaneously.

    Example:
        Basic Usage::

            from idasen import IdasenDesk


            async with IdasenDesk(mac="AA:AA:AA:AA:AA:AA") as desk:
                # call methods here...
    """

    #: Minimum desk height in meters.
    MIN_HEIGHT: float = 0.62

    #: Maximum desk height in meters.
    MAX_HEIGHT: float = 1.27

    #: Number of times to retry upon failure to connect.
    RETRY_COUNT: int = 3

    def __init__(self, mac: str, exit_on_fail: bool = False):
        self._logger = _DeskLoggingAdapter(
            logger=logging.getLogger(__name__), extra={"mac": mac}
        )
        self._mac = mac
        self._exit_on_fail = exit_on_fail
        self._client = BleakClient(self._mac)

    async def __aenter__(self):
        await self._connect()
        return self

    async def __aexit__(self, *args, **kwargs) -> Optional[bool]:
        return await self._client.__aexit__(*args, **kwargs)

    async def _connect(self):
        i = 0
        while True:
            try:
                await self._client.__aenter__()
                return
            except Exception:
                if i >= self.RETRY_COUNT:
                    self._logger.critical("Connection failed")
                    if self._exit_on_fail:
                        sys.exit(1)
                    raise
                i += 1
                self._logger.warning(
                    f"Failed to connect, retrying ({i}/{self.RETRY_COUNT})..."
                )
                time.sleep(0.3 * i)

    async def is_connected(self) -> bool:
        """
        Check connection status of the desk.

        Returns:
            Boolean representing connection status.

        >>> async def example() -> bool:
        ...     async with IdasenDesk(mac="AA:AA:AA:AA:AA:AA") as desk:
        ...         return await desk.is_connected()
        >>> asyncio.run(example())
        True
        """
        return await self._client.is_connected()

    @property
    def mac(self) -> str:
        """ Desk MAC address. """
        return self._mac

    async def move_up(self):
        """
        Move the desk upwards.

        This command moves the desk upwards for a fixed duration
        (approximately one second) as set by your desk controller.

        >>> async def example():
        ...     async with IdasenDesk(mac="AA:AA:AA:AA:AA:AA") as desk:
        ...         await desk.move_up()
        >>> asyncio.run(example())
        """
        await self._client.write_gatt_char(_UUID_COMMAND, _COMMAND_UP, response=False)

    async def move_down(self):
        """
        Move the desk downwards.

        This command moves the desk downwards for a fixed duration
        (approximately one second) as set by your desk controller.

        >>> async def example():
        ...     async with IdasenDesk(mac="AA:AA:AA:AA:AA:AA") as desk:
        ...         await desk.move_down()
        >>> asyncio.run(example())
        """
        await self._client.write_gatt_char(_UUID_COMMAND, _COMMAND_DOWN, response=False)

    async def move_to_target(self, target: float):
        """
        Move the desk to the target position.

        Args:
            target: Target position in meters.

        Raises:
            ValueError: Target exceeds maximum or minimum limits.

        >>> async def example():
        ...     async with IdasenDesk(mac="AA:AA:AA:AA:AA:AA") as desk:
        ...         await desk.move_to_target(1.1)
        >>> asyncio.run(example())
        """
        if target > self.MAX_HEIGHT:
            raise ValueError(
                f"target position of {target:.3f} meters exceeds maximum of "
                f"{self.MAX_HEIGHT:.3f}"
            )
        elif target < self.MIN_HEIGHT:
            raise ValueError(
                f"target position of {target:.3f} meters exceeds minimum of "
                f"{self.MIN_HEIGHT:.3f}"
            )

        while True:
            height = await self.get_height()
            difference = target - height
            self._logger.debug(f"{target=} {height=} {difference=}")
            if abs(difference) < 0.005:  # tolerance of 0.005 meters
                self._logger.info(f"reached target of {target:.3f}")
                await self.stop()
                return
            elif difference > 0:
                await self.move_up()
            elif difference < 0:
                await self.move_down()

    async def stop(self):
        """ Stop desk movement. """
        await asyncio.gather(
            self._client.write_gatt_char(_UUID_COMMAND, _COMMAND_STOP, response=False),
            self._client.write_gatt_char(
                _UUID_REFERENCE_INPUT, _COMMAND_REFERENCE_INPUT_STOP, response=False
            ),
        )

    async def get_height(self) -> float:
        """
        Get the desk height in meters.

        Returns:
            Desk height in meters.

        >>> async def example() -> float:
        ...     async with IdasenDesk(mac="AA:AA:AA:AA:AA:AA") as desk:
        ...         await desk.move_to_target(1.0)
        ...         return await desk.get_height()
        >>> asyncio.run(example())
        1.0
        """
        return _bytes_to_meters(await self._client.read_gatt_char(_UUID_HEIGHT))

    @classmethod
    async def discover(cls) -> Optional[str]:
        """
        Try to find the desk's MAC address by discovering currently connected devices.

        Returns:
            MAC address if found, ``None`` if not found.
        """
        try:
            devices = await discover()
        except Exception:
            return None
        return next(
            (device.address for device in devices if device.name.startswith("Desk")),
            None,
        )

idasen-ui/ui.py
# IDASEN UI - DESK CONTROL
import wx
import wx.adv
import wx.lib.agw.gradientbutton as GB
import wx.lib.agw.aquabutton as AB
import wx.lib.buttons as GBB
import functools

from idasen import IdasenDesk
from typing import Callable
from typing import List
from typing import Optional
from threading import *
import asyncio
import logging
import os
import sys
import voluptuous as vol
import yaml
import time

HOME = os.path.expanduser("~")
IDASEN_CONFIG_DIRECTORY = os.path.join(HOME, ".config", "idasen-ui")
IDASEN_CONFIG_PATH = os.path.join(IDASEN_CONFIG_DIRECTORY, "idasen-ui.yaml")
LOG_TO_CONSOLE = False

DEFAULT_CONFIG = {
    "positions": {"pos2": 1.1, "pos1": 0.70},
    "mac_address": "AA:AA:AA:AA:AA:AA",
}

CONFIG_SCHEMA = vol.Schema(
    {
        "mac_address": vol.All(str, vol.Length(min=17, max=17)),
        "positions": {
            str: vol.All(
                vol.Any(float, int),
                vol.Range(min=IdasenDesk.MIN_HEIGHT, max=IdasenDesk.MAX_HEIGHT),
            )
        },
    },
    extra=False,
)

       
def log(msg):
    if LOG_TO_CONSOLE:
        print(msg)
        
def align_bottom_right(win):
    dw, dh = wx.DisplaySize()
    w, h = win.GetSize()
    x = dw - w
    y = dh - h - 40
    win.SetPosition((x, y))
    
def message_to_user(msg):
        dlg = wx.MessageDialog(None, msg, "Message", wx.OK|wx.ICON_EXCLAMATION)
        dlg.ShowModal()
        dlg.Destroy()        
                     
                     
def save_config(config: dict, path: str = IDASEN_CONFIG_PATH):
    with open(path, "w") as f:
        yaml.dump(config, f)


def load_config(path: str = IDASEN_CONFIG_PATH) -> dict:
    """ Load user config. """
    try:
        with open(path, "r") as f:
            config = yaml.load(f, Loader=yaml.FullLoader)
    except FileNotFoundError:
        return {}

    try:
        config = CONFIG_SCHEMA(config)
    except vol.Invalid as e:
        log(f"Invalid configuration: {e}", file=sys.stderr)
        sys.exit(1)
        
    return config
        
async def discover_desk() -> bool:
    mac = await IdasenDesk.discover()
    global config
    if mac is not None:
        log(f"Discovered desk's MAC address: {mac}")
        if os.path.isfile(IDASEN_CONFIG_PATH):
            #update existing config
            config = load_config()
            config["mac_address"] = mac            
            save_config(config)
        else:
            #create new config file
            DEFAULT_CONFIG["mac_address"] = mac
            os.makedirs(IDASEN_CONFIG_DIRECTORY, exist_ok=True)            
            config = DEFAULT_CONFIG
            save_config(config)            
        return True
    else:
        return False
                
########################################################################
# Thread class that executes processing
class DeskWorkerThread(Thread):
    """Worker Thread Class."""
    def __init__(self, notify_window):
        """Init Worker Thread Class."""                
        Thread.__init__(self)
        self._notify_window = notify_window
        self.current_height = 0.0
        self.desk_height_target = 0.0
        self.workerThread = False

    def connect(self) -> bool:
        self.idasen_desk = IdasenDesk(config["mac_address"], exit_on_fail=False)          
        asyncio.run(self.idasen_desk._connect())            
        self.connected = asyncio.run(self.idasen_desk.is_connected())
        return self.connected    

    def start_running_loop(self):
        # This starts the thread running loop
        self.workerThread = True
        self.start()

    def stop_running_loop(self):
        self.workerThread = False        

    def is_connected(self) -> bool:
        return self.connected    

    def move_to_height(self, height):
        if height > self.idasen_desk.MAX_HEIGHT:
            log(f"target height of {height:.3f} meters exceeds maximum of {self.MAX_HEIGHT:.3f}")
            self.desk_height_target  = 0.0
        elif height < self.idasen_desk.MIN_HEIGHT:
            log(f"target height of {height:.3f} meters exceeds minimum of {self.MIN_HEIGHT:.3f}")
            self.desk_height_target  = 0.0    
        else:
            log(f"moving to target height of {height:.3f} meters")
            self.desk_height_target = height            

    def run(self):
        """Run Worker Thread."""   
        log("Starting worker thread...")

        deskMovingUp = False
        deskMovingDown = False
        previous_difference = 0.0
        bug_protection_counter = 0
        bug_protection_retry = 0
        refresh_counter_limit = 10
        refresh_auto_counter = refresh_counter_limit        

        while self.workerThread:
            # pseudo-realtime running loop, everything in there should be quick
            # move up sequence
            if self._notify_window.buttonUpPressed:
                log("moving up...")
                asyncio.run(self.idasen_desk.move_up())
                deskMovingUp = True
                deskMovingDown = False             
                self.desk_height_target = 0.0                
                refresh_auto_counter = refresh_counter_limit #force refresh
            # move down sequence
            elif self._notify_window.buttonDownPressed:
                log("moving down...")
                asyncio.run(self.idasen_desk.move_down())
                deskMovingUp = False
                deskMovingDown = True     
                self.desk_height_target = 0.0   
                refresh_auto_counter = refresh_counter_limit #force refresh
            # stop moving
            elif deskMovingUp or deskMovingDown:
                log("stop moving up...")
                asyncio.run(self.idasen_desk.stop())
                deskMovingUp = False
                deskMovingDown = False        
                self.desk_height_target = 0.0   
                refresh_auto_counter = refresh_counter_limit #force refresh
                
            # move_to_height button 1 or 2 pressed, let's move to target
            if self.desk_height_target != 0.0:
                difference = self.desk_height_target - self.current_height
                log(f"{self.desk_height_target=} {self.current_height=} {difference=}")  
                #------------------------------------------
                # ---- protection for idasen desk move issue
                # ---- but only retry twice to reach the target height
                # ---- required in case something blocks the desk from moving up or down
                if previous_difference == difference:
                    bug_protection_counter = bug_protection_counter + 1
                else:
                    bug_protection_counter = 0
                    bug_protection_retry = 0
                    previous_difference = difference                
                if bug_protection_counter > 9:
                    log("waiting 1 sec for desk to catch up...")
                    time.sleep(1)
                    bug_protection_retry = bug_protection_retry + 1
                if bug_protection_retry > 2:
                    log("Someting wrong... cancelling move_to_height")
                    asyncio.run(self.idasen_desk.stop())
                    deskMovingUp = False
                    deskMovingDown = False        
                    self.desk_height_target = 0.0   
                    refresh_auto_counter = refresh_counter_limit #force refresh                         
                # ---- end of protection                        
                #------------------------------------------
                if abs(difference) < 0.005:  # tolerance of 0.005 meters
                    log(f"reached target of {self.desk_height_target:.3f}")
                    self.desk_height_target = 0.0
                    asyncio.run(self.idasen_desk.stop())                   
                elif difference > 0:
                    log("moving up...")
                    asyncio.run(self.idasen_desk.move_up())  
                elif difference < 0:
                    log("moving down...")
                    asyncio.run(self.idasen_desk.move_down())                    
                refresh_auto_counter = refresh_counter_limit #force refresh                   

            #auto-refresh current height label
            if refresh_auto_counter >= refresh_counter_limit:                
                height = asyncio.run(self.idasen_desk.get_height())
                if self.current_height != height:
                    self.current_height = height
                    self._notify_window.gbHeightBtn.SetLabel(f"{self.current_height:.2f}")
                    self._notify_window.gbHeightBtn.Refresh() 
                refresh_auto_counter = 0
            else:
                # we are IDLE... refresh UI slowly
                refresh_auto_counter = refresh_auto_counter + 1
                time.sleep(0.5)

        # End of while loop
        log("Returning from worker thread.")

        
class MyForm(wx.Frame):
 
    #----------------------------------------------------------------------
    def __init__(self):
        size = wx.Size(465,90)
        style = wx.DEFAULT_FRAME_STYLE & ~(wx.RESIZE_BORDER | wx.MAXIMIZE_BOX| wx.SYSTEM_MENU  | wx.CLIP_CHILDREN)
        self.myFrame = wx.MiniFrame.__init__(self,None, wx.ID_ANY, "Idasen - Desk Control", wx.DefaultPosition, size, style, "")
                
        icon = wx.Icon()
        icon.CopyFromBitmap(wx.Bitmap("appicon.png", wx.BITMAP_TYPE_ANY))
        self.SetIcon(icon)        
        
        panel = wx.Panel(self, wx.ID_ANY)      
        
        bmp = wx.Bitmap("bt-nc.png", wx.BITMAP_TYPE_ANY)          
        btsize = wx.Size(60,46)
        self.gbBluetoothBtn = GB.GradientButton(panel, bitmap=bmp, label="", size=btsize)                        
        self.gbBluetoothBtn.Bind(wx.EVT_BUTTON, self.onBtBtnPress) 
        self.gbBluetoothBtn.SetToolTip(wx.ToolTip("Make sure Idasen desk is connected to computer.\nPress Bluetooth button to discover desk."))                         
                                  
        self.gbHeightBtn = GB.GradientButton(panel, label="N/A")
        size = wx.Size(75,46)
        font = wx.Font(wx.FontInfo(24).FaceName("Arial").Bold())
        self.gbHeightBtn.SetFont(font)
        self.gbHeightBtn.SetInitialSize(size)
        self.gbHeightBtn.Disable()
		
        bmp = wx.Bitmap("up-nc.png", wx.BITMAP_TYPE_ANY)
        self.gbUpBtn = GB.GradientButton(panel, bitmap=bmp, label="", size=btsize)        
        self.gbUpBtn.Bind( wx.EVT_LEFT_DOWN, self.onBtnUpPress);
        self.gbUpBtn.Bind( wx.EVT_LEFT_UP, self.onBtnUpRelease);
        self.gbUpBtn.Disable()
        
        bmp = wx.Bitmap("down-nc.png", wx.BITMAP_TYPE_ANY)
        self.gbDownBtn = GB.GradientButton(panel, bitmap=bmp, label="", size=btsize)
        self.gbDownBtn.Bind( wx.EVT_LEFT_DOWN, self.onBtnDownPress);
        self.gbDownBtn.Bind( wx.EVT_LEFT_UP, self.onBtnDownRelease);
        self.gbDownBtn.Disable()
        
        bmp = wx.Bitmap("pos1-nc.png", wx.BITMAP_TYPE_ANY)
        self.pos1Btn = GB.GradientButton(panel, bitmap=bmp, label="", size=btsize)
        self.pos1Btn.Bind(wx.EVT_BUTTON, self.onBtn1Press)  
        self.pos1Btn.Disable()       

        bmp = wx.Bitmap("pos2-nc.png", wx.BITMAP_TYPE_ANY)
        self.pos2Btn = GB.GradientButton(panel, bitmap=bmp, label="", size=btsize)
        self.pos2Btn.Bind(wx.EVT_BUTTON, self.onBtn2Press)   
        self.pos2Btn.Disable()       

        bmp = wx.Bitmap("m-nc.png", wx.BITMAP_TYPE_ANY)
        self.gbMBtn = GB.GradientButton(panel, bitmap=bmp, label="", size=btsize)
        self.gbMBtn.Bind(wx.EVT_BUTTON, self.onBtnMemoryPress)
        self.gbMBtn.Disable()
	
        # Create desk instance that will be running in a separate thread        
        self.buttonUpPressed = False
        self.buttonDownPressed = False    
        self.buttonMemoryPressed = False
        self.idasen_desk = DeskWorkerThread(self) 
        # Try to connect to Idasen desk based on previous saved config
        try:            
            if self.idasen_desk.connect():            
                self.showConnectedButton()
                self.idasen_desk.start_running_loop()
        
        except Exception as e:
            log("No saved config found")
            
            
        sizer = wx.BoxSizer(wx.HORIZONTAL)
        sizer.Add(self.gbBluetoothBtn, 0, wx.ALL, 1)
        sizer.Add(self.gbHeightBtn, 0, wx.ALL, 1)
        sizer.Add(self.gbUpBtn, 0, wx.ALL, 1)
        sizer.Add(self.gbDownBtn, 0, wx.ALL, 1)
        sizer.Add(self.pos1Btn, 0, wx.ALL, 1)
        sizer.Add(self.pos2Btn, 0, wx.ALL, 1)
        sizer.Add(self.gbMBtn, 0, wx.ALL, 1)
        
        panel.SetSizer(sizer)

            
        self.Bind(wx.EVT_CLOSE, self.OnClose)
    #----------------------------------------------------------------------
    def OnClose(self, event):
        self.idasen_desk.stop_running_loop()
        time.sleep(0.5) #wait for thread to exit
        event.Skip()

    def showConnectedButton(self):        
        self.gbBluetoothBtn.SetBitmapLabel(wx.Bitmap("bt.png", wx.BITMAP_TYPE_ANY))
        self.gbBluetoothBtn.Disable()
        self.gbUpBtn.SetBitmapLabel(wx.Bitmap("up.png", wx.BITMAP_TYPE_ANY))        
        self.gbUpBtn.Enable()        
        self.gbDownBtn.SetBitmapLabel(wx.Bitmap("down.png", wx.BITMAP_TYPE_ANY))
        self.gbDownBtn.Enable()        
        self.pos1Btn.SetBitmapLabel(wx.Bitmap("pos1.png", wx.BITMAP_TYPE_ANY))
        self.pos1Btn.Enable()               
        self.pos2Btn.SetBitmapLabel(wx.Bitmap("pos2.png", wx.BITMAP_TYPE_ANY))
        self.pos2Btn.Enable()               
        self.gbMBtn.SetBitmapLabel(wx.Bitmap("m.png", wx.BITMAP_TYPE_ANY))
        self.gbMBtn.Enable()
        

    
    def onBtBtnPress(self, event):
        """"""        
        log("BT button pressed! Trying to discover_desk...")
        if asyncio.run(discover_desk()):
            log("Desk found, trying to connect...")
            if self.idasen_desk.connect():
                log("Desk connected! Enabling and starting running loop...")
                self.showConnectedButton()
                self.idasen_desk.start_running_loop()

    def onBtnUpPress(self, event):
        """"""
        self.buttonUpPressed = True
        
    def onBtnUpRelease(self, event):
        """"""
        self.buttonUpPressed = False
        
    def onBtnDownPress(self, event):
        """"""
        self.buttonDownPressed = True
        
    def onBtnDownRelease(self, event):
        """"""
        self.buttonDownPressed = False  

    def onBtn1Press(self, event):
        """"""
        if self.buttonMemoryPressed:
            self.disableSavePosition()
            self.saveCurrentHeightInConfig("pos1")
        else:        
            self.idasen_desk.move_to_height(config["positions"]["pos1"])
        
        
    def onBtn2Press(self, event):
        """"""
        if self.buttonMemoryPressed:
            self.disableSavePosition()
            self.saveCurrentHeightInConfig("pos2")
        else:
            self.idasen_desk.move_to_height(config["positions"]["pos2"])
        
        
    def onBtnMemoryPress(self, event):
        """"""
        if self.buttonMemoryPressed:
            self.disableSavePosition()
        else:
            self.buttonMemoryPressed = True
            bmp = wx.Bitmap("pos1-h.png", wx.BITMAP_TYPE_ANY)
            self.pos1Btn.SetBitmapLabel(bmp)
            bmp = wx.Bitmap("pos2-h.png", wx.BITMAP_TYPE_ANY)
            self.pos2Btn.SetBitmapLabel(bmp)
            self.gbUpBtn.Disable()
            self.gbDownBtn.Disable()        

    def saveCurrentHeightInConfig(self,savePos):
        config = load_config()
        config["positions"][savePos] = self.idasen_desk.current_height
        save_config(config)
    
    def disableSavePosition(self):
        self.buttonMemoryPressed = False
        bmp = wx.Bitmap("pos1.png", wx.BITMAP_TYPE_ANY)
        self.pos1Btn.SetBitmapLabel(bmp)
        bmp = wx.Bitmap("pos2.png", wx.BITMAP_TYPE_ANY)
        self.pos2Btn.SetBitmapLabel(bmp)
        self.gbUpBtn.Enable()
        self.gbDownBtn.Enable()        

   
# Run the program
config = load_config()
if __name__ == "__main__":   
    app = wx.App(False)
    frame = MyForm()
    align_bottom_right(frame)    
    frame.Show()
    app.MainLoop()
    